File: 2022/03/java/Solution.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Solution {

	public static void main(String[] args) throws IOException {
		FileReader fileReader = new FileReader("../input.txt");
		BufferedReader bufferedReader = new BufferedReader(fileReader);

		String line;
		List<String> input = new ArrayList<>();

		while ((line = bufferedReader.readLine()) != null) {
			input.add(line);
		}

		bufferedReader.close();

		System.out.println("Part 1: " + part_1(input));
		System.out.println("Part 2: " + part_2(input));
	}

	public static int part_1(List<String> input) {
		int sum = 0;

		for (String line : input) {
			String first_half = line.substring(0, line.length() / 2);
			String second_half = line.substring(line.length() / 2);

			for (int c : first_half.chars().toArray()) {
				if (second_half.contains("" + ((char) c))) {
					if (c >= 'a') {
						sum += c - 'a' + 1;
					} else {
						sum += c - 'A' + 27;
					}
					break;
				}
			}
		}
		return sum;
	}

	public static int part_2(List<String> input) {

		int sum = 0;

		Iterator<String> iterator = input.iterator();

		while (iterator.hasNext()) {
			String a = iterator.next(), b = iterator.next(), c = iterator.next();
			List<Integer> a_b = new ArrayList<>();

			for (int ch : a.chars().toArray()) {
				if (b.contains("" + ((char) ch))) {
					a_b.add(ch);
				}
			}

			for (int ch : c.chars().toArray()) {
				if (a_b.contains(ch)) {
					if (ch >= 'a') {
						sum += ch - 'a' + 1;
					} else {
						sum += ch - 'A' + 27;
					}
					break;
				}
			}

		}

		return sum;
	}
}

File: 2022/03/cpp/solution.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

int part_1(std::vector<std::string> &lines) {
  int sum = 0;

  for (int i = 0; i < lines.size(); i++) {
    std::string line = lines[i];

    for (int j = 0; j < line.size() / 2; j++) {
      bool found = false;
      for (int k = line.size() / 2; k < line.size(); k++) {
        if (line[j] == line[k]) {
          found = true;
          break;
        }
      }
      if (found) {
        if (line[j] >= 'a') {
          sum += line[j] - 'a' + 1;
        } else {
          sum += line[j] - 'A' + 27;
        }
        break;
      }
    }
  }

  return sum;
}

int part_2(std::vector<std::string> &lines) {
  int sum = 0;

  for (int i = 0; i < lines.size(); i += 3) {
    for (int a = 0; a < lines[i].size(); a++) {
      bool found = false;
      for (int b = 0; b < lines[i + 1].size(); b++) {
        if (lines[i][a] == lines[i + 1][b]) {
          for (int c = 0; c < lines[i + 2].size(); c++) {
            if (lines[i][a] == lines[i + 2][c]) {
              found = true;
              break;
            }
          }
        }

        if (found) {
          break;
        }
      }

      if (found) {
        if (lines[i][a] >= 'a') {
          sum += lines[i][a] - 'a' + 1;
        } else {
          sum += lines[i][a] - 'A' + 27;
        }
        break;
      }
    }
  }

  return sum;
}

int main() {

  std::vector<std::string> lines;
  std::ifstream file("../input.txt");

  std::string line;
  while (std::getline(file, line)) {
    lines.push_back(line);
  }

	file.close();

  std::cout << "Part 1: " << part_1(lines) << std::endl;
  std::cout << "Part 2: " << part_2(lines) << std::endl;

  return 0;
}
